paddle/fluid/operators/optimizers/adam_op.h

```diff
@@ -506,21 +508,58 @@ class AdamOpKernel : public framework::OpKernel<T> {
                           beta2_pow_out->numel()));
 
     if (grad_var->IsType<framework::LoDTensor>()) {
-      auto* grad = ctx.Input<LoDTensor>("Grad");
+      T beta1_p = beta1_pow->data<T>()[0];
+      T beta2_p = beta2_pow->data<T>()[0];
 
-      AdamFunctor<T, CPUAdam> functor(
-          beta1, beta2, epsilon, beta1_pow->data<T>(), beta2_pow->data<T>(),
-          mom1->data<T>(), mom1_out->mutable_data<T>(ctx.GetPlace()),
-          mom2->data<T>(), mom2_out->mutable_data<T>(ctx.GetPlace()),
-          lr->data<T>(), grad->data<T>(), param->data<T>(),
-          param_out->mutable_data<T>(ctx.GetPlace()));
-      functor(param->numel());
       if (!use_global_beta_pow) {
         beta1_pow_out->mutable_data<T>(ctx.GetPlace())[0] =
             beta1 * beta1_pow->data<T>()[0];
         beta2_pow_out->mutable_data<T>(ctx.GetPlace())[0] =
             beta2 * beta2_pow->data<T>()[0];
       }
+
+      auto* grad = ctx.Input<LoDTensor>("Grad");
+
+      T* param_out_ptr = param_out->mutable_data<T>(ctx.GetPlace());
+      T* mom1_out_ptr = mom1_out->mutable_data<T>(ctx.GetPlace());
+      T* mom2_out_ptr = mom2_out->mutable_data<T>(ctx.GetPlace());
+
+      T learning_rate = lr->data<T>()[0] * (sqrt(1 - beta2_p) / (1 - beta1_p));
+      T eps = epsilon * sqrt(1 - beta2_p);
+
+      jit::adam_attr_t attr(beta1, beta2);
+      int64_t numel = param->numel();
+
+      const T* param_ptr = param->data<T>();
+      const T* mom1_ptr = mom1->data<T>();
+      const T* mom2_ptr = mom2->data<T>();
+      const T* grad_ptr = grad->data<T>();
+
+      auto adam =
+          jit::KernelFuncs<jit::AdamTuple<T>, platform::CPUPlace>::Cache().At(
+              attr);
+
+      static constexpr int64_t chunk_size = 512;
+
+#ifdef PADDLE_WITH_MKLML
+#pragma omp parallel for
```

Comment:
Just a suggesstion: I would suggest decomposing "#pragma omp parallel for" into:
#pragma parallel       
   #pragma omp nowait
   // major full block computation here
   #pragma single nowait
  // tail computation here

Reply:
Very nice idea thanks! As we discussed offline, I'll implement that soon in another PR